### PR TITLE
ci-operator: optionally use build cache for the root

### PR DESCRIPTION
**Pull request by stevekuznetsov**
ci-operator: validate that build_root ISTags are fully formed

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

ci-operator: clean up build root step generation

There's no place where any of the weird "test-base" stuff is used and
the previous commit validates as much. I'm not sure what it was for and
it's not documented anywhere. I removed that handler and fixed the git
base / project build options to pass through the full struct, instead of
cherry-picking some options and silently dropping the rest.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

ci-operator: optionally use build cache for the root

When the user asks us to, we can determine if the build cache we have
published for the repo is up-to-date. If it is, we should use it for our
build root instead of the build root that would otherwise be used. In
cases where we can't resolve it or the cache is stale, fall back to the
original root.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/assign @petr-muller @alvaroaleman 

need to think about how this might be e2e tested or it's worth it

### CODE REVIEW

**Comment by alvaroaleman** [comment]
@stevekuznetsov can this be split up into "Improvements to existing logic" and "Add support to use a build cache"? Can you also write three sentences high level description of how the build cache is intended to works?

File: pkg/defaults/defaults.go
Comment (petr-muller) [inline pkg/defaults/defaults.go:420]: Maybe we could be robust against random flakes here and just use the normal root. If the cluster is in serious trouble than something will fail later, but cache flakes should not fail a job.

File: pkg/defaults/defaults.go
Comment (petr-muller) [inline pkg/defaults/defaults.go:428]: typo: chace->cache

File: pkg/defaults/defaults.go
Comment (petr-muller) [inline pkg/defaults/defaults.go:425]: Same here -> Is there a way how we easily discover problems like this without failing jobs for the user whose job would be fine if we didn't try to be smart?

**Comment by stevekuznetsov** [comment]
@alvaroaleman broke out the tech debt into https://github.com/openshift/ci-tools/pull/1811

I'll add docs to the site on the build-root later but the basic idea is:

1. publish `bin` for repositories, labeling it with the version of `root` used to build it
2. use this published image instead of `root` when the published image is based on the same `root`

Ideally, this means that:

1. `clonerefs` runs in a pre-populated repo from the last push and only does incremental fetches
2. all of the `go` build cache exists as it needs to and we get incremental builds
3. any changes to the `root` will propagate as needed, allowing people to seamlessly update their root without caring about invalidating the cache on purpose

**Review by petr-muller — COMMENTED** [review]
I was also thinking if there are some corner cases when people change their `build_root` configs but I was not able to come up with an idea of what could go wrong so it's probably fine :shrug: 

How does this work in the multi-cluster environment? Do I understand correctly that the `bin` image gets promoted in the standard promotion job (after each merge) which gets the image to the `app.ci` registry. What then distributes these images to build farms?

**Comment by stevekuznetsov** [comment]
@petr-muller 

> I was also thinking if there are some corner cases when people change

I don't think this is possible as we will only ever use the cache when it's based on the same root image digest as `root` would otherwise have been for that run.

> Do I understand correctly that

Yes.

> What then distributes these images to build farms?

That's what the `imageextraction.go` changes are for - that's the code that determines which images are required to be distributed across clusters for a given config.

File: pkg/defaults/defaults.go
Comment (stevekuznetsov) [reply pkg/defaults/defaults.go:420]: I thought about that, but I'm not sure that there's any difference between failing on a GET ISTag call for the cache or the root - both are against the same API server, so /shrug

File: pkg/defaults/defaults.go
Comment (stevekuznetsov) [reply pkg/defaults/defaults.go:425]: I thought about making this entirely best-effort but yes as you note it means that failures will be silent and hard to detect.

I also don't know that this really increases the burden on the job for success. We have three things that can fail here:

1. two GET ISTag calls
2. a JSON unmarshal

One of the two GET calls are calls we'd be doing anyway when we resolve `root`, so we only add the cost of potentially flaking here. I don't know that we really see flakes in the API server for our clusters, they are either up or not.

The JSON unmarhsal can never fail in a working system. I don't think it's possible for the build API to produce bad metadata here in the first place.

IMHO as-is gives us the best ratio of usefulness/discoverability/etc.

**Comment by stevekuznetsov** [comment]
/hold

For https://github.com/openshift/ci-tools/pull/1811

File: pkg/api/helper/imageextraction.go
Comment (stevekuznetsov) [inline pkg/api/helper/imageextraction.go:38]: We need to land this to test it in rehearsal even :|

File: pkg/defaults/defaults.go
Comment (petr-muller) [reply pkg/defaults/defaults.go:427]: Makes sense!

**Review by petr-muller — APPROVED** [review]

**Comment by stevekuznetsov** [comment]
Just rebased to remove the first two commits.

**Pull request merged**
